streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -854,16 +870,216 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
                                                                                  final JoinWindows windows,
                                                                                  final Serde<K> keySerde,
                                                                                  final Serde<V> valueSerde) {
-        return Stores.windowStoreBuilder(
-            Stores.persistentWindowStore(
-                joinName + "-store",
-                Duration.ofMillis(windows.size() + windows.gracePeriodMs()),
-                Duration.ofMillis(windows.size()),
-                true
-            ),
-            keySerde,
-            valueSerde
+        final WindowBytesStoreSupplier supplier = Stores.persistentWindowWithTimestampStore(
+            joinName + "-store",
+            Duration.ofMillis(windows.size() + windows.gracePeriodMs()),
+            Duration.ofMillis(windows.size()),
+            true
         );
+        return new WindowStoreBuilder<K, V>(supplier, null, null, Time.SYSTEM) {
+            final StoreBuilder<WindowStore<K, ValueAndTimestamp<V>>> inner = Stores.windowWithTimestampStoreBuilder(
+                supplier,
+                keySerde,
+                valueSerde
+            );
+
+            @Override
+            public WindowStoreBuilder<K, V> withCachingEnabled() {
+                inner.withCachingEnabled();
+                return this;
+            }
+
+            @Override
+            public WindowStoreBuilder<K, V> withCachingDisabled() {
+                inner.withCachingDisabled();
+                return this;
+            }
+
+            @Override
+            public WindowStoreBuilder<K, V> withLoggingEnabled(final Map<String, String> config) {
+                inner.withLoggingEnabled(config);
+                return this;
+            }
+
+            @Override
+            public WindowStoreBuilder<K, V> withLoggingDisabled() {
+                inner.withLoggingDisabled();
+                return this;
+            }
+
+            @Override
+            public Map<String, String> logConfig() {
+                return inner.logConfig();
+            }
+
+            @Override
+            public boolean loggingEnabled() {
+                return inner.loggingEnabled();
+            }
+
+            @Override
+            public String name() {
+                return inner.name();
+            }
+
+            @Override
+            public WindowStore<K, V> build() {
+                return new WindowStoreFacade<>(inner.build());
+            }
+        };
+    }
+
+    public static class WindowStoreFacade<A, B> implements WindowStore<A, B>, RecordConverter {
```

Comment:
QQ:

1. Does this class need to be a public static one?
2. Can it extend form `ReadOnlyWindowStoreFacade`?

Reply:
I guess in the end we will find these classes a better place (currently they are a bit scattered, e.g. `KeyValueStoreFacade` is in the materializer class).